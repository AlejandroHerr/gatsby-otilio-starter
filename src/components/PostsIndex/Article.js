import React from 'react';
import Link from 'gatsby-link';

import { articleType } from '../../propTypes/postsIndex';

import styles from './PostsIndex.module.scss';

const getDisplayDate = date => date.toLocaleDateString('en-us', { day: 'numeric', month: 'short', year: 'numeric' });

const Article = ({ article }) => {
  const {
    author,
    date,
    excerpt,
    slug,
    tags,
    title,
  } = article;

  const postDate = new Date(date);

  return (
    <article className={styles.posts_index_article}>
      <div className={styles.posts_index_article__inner}>
        <header className="post-header">
          <h2 className={styles.posts_index_article__title}>
            <Link
              className={styles.posts_index_article__title__link}
              to={slug}
            >
              {title
              }
            </Link>
          </h2>
          <span className={styles.posts_index_article__meta}>
            {author}
            {tags && ` on ${tags.join(', ')}`}
            {' | '}
            <time dateTime={postDate.toISOString()}>{getDisplayDate(postDate)}</time>
          </span>
          <div className={styles.posts_index_article__clear} />
        </header>
        <section>
          <p className={styles.posts_index_article__excerpt}>{excerpt}</p>
        </section>
      </div>
      <div className={styles.posts_index_article__separator} />
    </article>
  );
};

Article.propTypes = {
  article: articleType.isRequired,
};

export default Article;
