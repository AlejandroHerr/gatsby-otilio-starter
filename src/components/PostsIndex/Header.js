// @flow
import React from 'react';
import classnames from 'classnames';

import type { SizeImageSharpType } from '../../types/gatsby';

import HeroHeader from '../HeroHeader';

import styles from './PostsIndex.module.scss';

type PropsType = {
  description: string,
  image: SizeImageSharpType,
  title: string,
};

const MainHeader = ({
  description, image, title,
}: PropsType) => (
  <HeroHeader
    className={classnames(styles.posts_index_header, {
        [styles['posts_index_header--has_image']]: !!image,
      })}
    image={image}
  >
    <h1
      className={classnames(styles.posts_index_header__title, {
          [styles['posts_index_header__title--has_image']]: !!image,
        })}
    >
      {title}
    </h1>
    <span
      className={classnames(styles.posts_index_header__body, {
          [styles['posts_index_header__body--has_image']]: !!image,
        })}
    >
      {description}
    </span>
  </HeroHeader>
);

export default MainHeader;
