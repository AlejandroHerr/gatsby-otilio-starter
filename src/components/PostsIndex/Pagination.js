// @flow
import React from 'react';
import Link from 'gatsby-link';
import classnames from 'classnames';

import type { PaginationType } from '../../types/postsIndex';

import ByteIcon from '../icons/ByteIcon';

import styles from './PostsIndex.module.scss';

const Pagination = ({ page, pageCount, pathPrefix }: PaginationType) => (
  <nav className={styles.posts_index_pagination}>
    <div className={styles.posts_index_pagination__inner}>
      {page > 1 &&
        <Link
          className={classnames(
            styles.posts_index_pagination__link,
            styles.posts_index_pagination__next,
          )}
          to={page > 2 ? `${pathPrefix}/${page - 1}` : '/'}
        >
          <span className={styles.posts_index_pagination__label}><ByteIcon icon="chevron_left" /> Newer Posts</span>
        </Link>
      }
      <span className={styles.posts_index_pagination__info}>{`Page ${page} of ${pageCount}`}</span>
      {page < pageCount &&
        <Link
          className={classnames(
            styles.posts_index_pagination__link,
            styles.posts_index_pagination__prev,
          )}
          to={`${pathPrefix}/${page + 1}`}
        >
          <span className={styles.posts_index_pagination__label}>Older Posts <ByteIcon icon="chevron_right" /></span>
        </Link>
      }
      <div className={styles.posts_index_pagination__clear} />
      <div className={styles.posts_index_pagination__separator} />
    </div>
  </nav>
);

export default Pagination;
