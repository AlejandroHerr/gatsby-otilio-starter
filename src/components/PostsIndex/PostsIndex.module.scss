@import '../../colors';
@import '../../fonts';

.posts_index_article {
  position: relative;
  padding: 4em 0;
  z-index: 10;
}

.posts_index_article__inner,
.posts_index_pagination__inner {
  max-width: 32em;
  margin: 0 auto;
  padding: 0 2em;
}

.posts_index_article__title {
  position: relative;
  margin: 0 0 .25em;
  color: $color-black;
  font-size: 2em;
  font-weight: 700;
  line-height: 1em;
  text-indent: -1px;
  z-index: 10;
}

.posts_index_article__title__link {
  color: inherit;
  font-weight: inherit;
  text-decoration: none;
}

.posts_index_article__meta {
  display: block;
  position: relative;
  margin-bottom: 1.334em;
  font-family: $header-font-familiy;
  font-size: .75em;
  font-weight: 400;
  line-height: 1.334em;
  opacity: .5;
  z-index: 10;
}

.posts_index_article__meta__link {
  border-bottom: 1px solid transparent;
  color: inherit;
  text-decoration: none;
}

.posts_index_article__meta__link:hover,
.posts_index_article__meta__link:focus {
  border-color: rgba(0, 0, 0, .5);
}

.posts_index_article__clear,
.posts_index_pagination__clear {
  clear: both;
}

.posts_index_article__excerpt {
  margin: .5em 0 0;
}

.posts_index_article__separator,
.posts_index_pagination__separator, {
  position: absolute;
  bottom: 0;
  left: 50%;
  width: 32em;
  height: 1px;
  margin-left: -16em;
  background: $color-gray-90;
}

@media only screen and (max-width: 480px) {
  .posts_index_article__inner,
  .posts_index_pagination__inner {
    padding: 0 16px;
  }
}

.posts_index_pagination {
  position: relative;
  height: 2em;
  margin: 0;
  padding: 2em 0;
  font-family: $header-font-familiy;
}

.posts_index_pagination__info {
  display: block;
  position: absolute;
  left: 0;
  width: 100%;
  height: 2.667em;
  font-size: .75em;
  line-height: 2.667em;
  text-align: center;
  z-index: 50;
}

.posts_index_pagination__link {
  font-size: .75em;
  z-index: 100;
  /*position: relative;
  display: block;
  height: 1.334em;
  margin: -1px 0;
  padding: .667em 1em;
  z-index: 100;
  border: 1px solid $color-blue-80;
  font-size: .75em;
  line-height: 1.5em;
  text-align: center;
  text-decoration: none;
  color: $color-blue-60;
  transition: all ease .3s;
  border-radius: 2em;*/
}

.posts_index_pagination__prev {
  float: right;
  padding-left: 1.334em;
}
.posts_index_pagination__next {
  float: left;
  padding-right: 1.334em;
}

.posts_index_pagination__label {
  display: inline-block;
  position: relative;
  top: -.167em;
  height: 1.334em;
  line-height: 1.334em;
}

.pagination a:hover,
.pagination a:focus {
	color: $primary-color;
	border-color: $primary-color;
}
.pagination a i {
	position: relative;
	width: auto;
	margin: auto;
	color: inherit;
	font-size: 1.334em;
	line-height: .75em;
}

.extra-pagination {
	display: none;
}
.extra-pagination .pagination:after {
	top: auto;
	bottom: 0;
}
.paged .extra-pagination {
	display: block;
}

@media only screen and (max-width: 320px) {
	.pagination .pagination-label {
		display: none;
	}
}
