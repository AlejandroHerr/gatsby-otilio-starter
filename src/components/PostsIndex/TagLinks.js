// @flow
import React from 'react';
import Link from 'gatsby-link';

import type { TagType } from '../../types/postsIndex';

type PropsType = {
  tags: Array<TagType>,
};

const TagLinks = ({ tags }: PropsType) => (
  <span>{' on '}{tags.map(({ tag, tagSlug }, idx) => (idx < (tags.length - 1)
  ? (
    <span key={tagSlug}>
      <Link to={`/tags/${tagSlug}`}>{tag}</Link>{', '}
    </span>
  )
  : (
    <span key={tag}>
      <Link to={`/tags/${tagSlug}`}>{tag}</Link>
    </span>
  )))}
  </span>
);

export default TagLinks;
