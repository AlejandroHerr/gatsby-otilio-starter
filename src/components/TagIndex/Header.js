// @flow
import React from 'react';
import classnames from 'classnames';

import type { SizeImageSharpType } from '../../types/gatsby';

import HeroHeader from '../HeroHeader';

import styles from './TagIndex.module.scss';

type PropsType = {
  image?: ?SizeImageSharpType,
  title: string,
  total: number,
};

const MainHeader = ({
  image, title, total,
}: PropsType) => (
  <HeroHeader
    className={classnames(styles.tag_index_header, {
      [styles['tag_index_header--has_image']]: !!image,
    })}
    image={image}
  >
    <h1
      className={classnames(styles.tag_index_header__title, {
          [styles['tag_index_header__title--has_image']]: !!image,
        })}
    >
      {title}
    </h1>
    <span
      className={classnames(styles.tag_index_header__body, {
          [styles['tag_index_header__body--has_image']]: !!image,
        })}
    >
      {`${total} posts in this category`}
    </span>
  </HeroHeader>
);


MainHeader.defaultProps = {
  image: null,
};

export default MainHeader;
