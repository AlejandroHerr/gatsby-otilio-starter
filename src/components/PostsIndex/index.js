// @flow
import PostsIndex from './PostsIndex';

export default PostsIndex;
