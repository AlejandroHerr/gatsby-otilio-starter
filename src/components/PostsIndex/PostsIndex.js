// @flow
import React from 'react';

import type { PathContextType, SiteInfoType } from '../../types/postsIndex';

import Header from './Header';
import Article from './Article';
import Pagination from './Pagination';

type PropsType = {
  pathContext: PathContextType,
  siteInfo: SiteInfoType,
};

const PostsIndex = ({ siteInfo, pathContext }: PropsType) => {
  const {
    group: posts,
    first,
    last,
    index,
    pageCount,
    pathPrefix,
  } = pathContext;

  return (
    <div>
      <Header
        image={siteInfo.cover.childImageSharp.sizes}
        title={siteInfo.title}
        description={siteInfo.shortDescription}
      />
      <main role="main">
        {posts.map(article => (
          <Article
            key={article.slug}
            article={article}
          />
        ))}
        <Pagination
          first={first}
          last={last}
          index={index}
          pageCount={pageCount}
          pathPrefix={pathPrefix}
        />
      </main>
    </div>
  );
};

export default PostsIndex;
