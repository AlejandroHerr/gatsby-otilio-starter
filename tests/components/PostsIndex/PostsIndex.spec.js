import React from 'react';
import { shallow } from 'enzyme';

import PostsIndex from '../../../src/components/PostsIndex/';
import Header from '../../../src/components/PostsIndex/Header';
import Article from '../../../src/components/PostsIndex/Article';
import Pagination from '../../../src/components/PostsIndex/Pagination';

const defaultProps = {
  pathContext: {
    group: [{
      slug: 'slug1',
    },
    {
      slug: 'slug2',
    }],
    first: true,
    last: false,
    index: 1,
    pageCount: 2,
    pathPrefix: 'prefix',
  },
  siteInfo: {
    shortDescription: 'Test Description',
    cover: {
      childImageSharp: {
        sizes: {
          src: 'testSrc',
        },
      },
    },
    title: 'Test Title',
    url: 'Test Url',
  },
};

const setup = (props = {}) => {
  const postsIndex = shallow(<PostsIndex {...defaultProps} {...props} />);

  return {
    postsIndex,
    header: postsIndex.find(Header),
    article: postsIndex.find(Article),
    pagination: postsIndex.find(Pagination),
  };
};

describe('PostsIndex', () => {
  it('should render a PostsIndex', () => {
    const {
      postsIndex, header, article, pagination,
    } = setup();

    expect(header).toHaveLength(1);
    expect(article).toHaveLength(defaultProps.pathContext.group.length);
    expect(pagination).toHaveLength(1);
    expect(postsIndex).toMatchSnapshot();
  });
});
