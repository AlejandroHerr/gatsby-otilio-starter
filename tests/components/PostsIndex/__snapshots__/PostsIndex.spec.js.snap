// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostsIndex should render a PostsIndex 1`] = `
<div>
  <MainHeader
    description="Test Description"
    image={
      Object {
        "src": "testSrc",
      }
    }
    title="Test Title"
    url="Test Url"
  />
  <main
    role="main"
  >
    <Article
      article={
        Object {
          "slug": "slug1",
        }
      }
      key="slug1"
    />
    <Article
      article={
        Object {
          "slug": "slug2",
        }
      }
      key="slug2"
    />
    <Pagination
      page={1}
      pageCount={2}
      pathPrefix="prefix"
    />
  </main>
</div>
`;
